### PR TITLE
Expand the smoke_test to validate key derivation and certificate signatures

### PR DESCRIPTION
**Pull request by korran**
As usual, I recommend reviewing each of the 14 commits independently.

### CODE REVIEW

File: sw-emulator/lib/periph/src/doe.rs [hunk 1]
```diff
@@ -410,4 +413,26 @@ mod tests {
         assert_eq!(soc_reg.doe_key(), expected_doe_key);
         assert_eq!(soc_reg.field_entropy(), expected_fe);
     }
+
+    #[test]
+    fn test_cbc() {
```
Comment (jhand2) [inline sw-emulator/lib/periph/src/doe.rs:418]: How is this related to cert testing?

File: test/tests/smoke_test.rs [hunk 2]
```diff
@@ -50,4 +99,53 @@ fn smoke_test() {
...
+    let ldev_cert = openssl::x509::X509::from_der(&ldev_cert_der).unwrap();
+    let ldev_cert_txt = String::from_utf8(ldev_cert.to_text().unwrap()).unwrap();
+
+    // To update the ldev cert testdata:
+    // std::fs::write("tests/smoke_testdata/ldevid_cert.txt", &ldev_cert_txt).unwrap();
+    // std::fs::write("tests/smoke_testdata/ldevid_cert.der", &ldev_cert_der).unwrap();
+
+    assert_eq!(
+        ldev_cert_txt.as_str(),
+        include_str!("smoke_testdata/ldevid_cert.txt")
```
Comment (jhand2) [inline test/tests/smoke_test.rs:119]: What's the rationale for comparing against both the der and text?

File: sw-emulator/lib/periph/src/doe.rs (hunk 1 above)
Comment (jhand2) [reply sw-emulator/lib/periph/src/doe.rs:418]: Oh, I just realized this is a PR into your own branch. Never mind :)

File: test/tests/smoke_test.rs [hunk 3]
```diff
@@ -50,4 +99,53 @@ fn smoke_test() {
...
+    let ldev_cert_der = hw
+        .mailbox_execute(TEST_ONLY_GET_LDEV_CERT, &[])
+        .unwrap()
+        .unwrap();
+    let ldev_cert = openssl::x509::X509::from_der(&ldev_cert_der).unwrap();
+    let ldev_cert_txt = String::from_utf8(ldev_cert.to_text().unwrap()).unwrap();
+
+    // To update the ldev cert testdata:
+    // std::fs::write("tests/smoke_testdata/ldevid_cert.txt", &ldev_cert_txt).unwrap();
+    // std::fs::write("tests/smoke_testdata/ldevid_cert.der", &ldev_cert_der).unwrap();
```
Comment (fdamato) [inline test/tests/smoke_test.rs:115]: is this test basically validating the DUT against  n-1 DUT ?

File: test/tests/smoke_test.rs (hunk 2 above)
Comment (korran) [reply test/tests/smoke_test.rs:119]: I'm not sure whether openssl includes every bit of the cert in the text, so this way we have both a human-readable diff in any PR that changes this logic, plus the actual bits for exhaustive assertion.

File: test/tests/smoke_test.rs (hunk 3 above)
Comment (korran) [reply test/tests/smoke_test.rs:115]: Yes, where n is a git commit. These certs and CSRs are generally not expected to change, so we want to make sure the test makes some noise if they do unexpectedly.

File: sw-emulator/lib/periph/src/doe.rs (hunk 1 above)
Comment (korran) [reply sw-emulator/lib/periph/src/doe.rs]: I've since rebased onto main (since that other PR has been submitted).

File: test/tests/smoke_test.rs (hunk 2 above)
Comment (jhand2) [reply test/tests/smoke_test.rs:119]: SGTM

File: test/tests/smoke_test.rs [hunk 4]
```diff
@@ -50,4 +99,53 @@ fn smoke_test() {
...
+
+    println!("ldev-cert: {}", ldev_cert_txt);
+
+    let fmc_alias_cert = hw
+        .mailbox_execute(TEST_ONLY_GET_FMC_ALIAS_CERT, &[])
+        .unwrap()
+        .unwrap();
+    let fmc_alias_cert = openssl::x509::X509::from_der(&fmc_alias_cert).unwrap();
+
+    // TODO: (DO NOT SUBMIT) Validate the contents / measurements in the fmc_alias_cert.
```
Comment (jhand2) [inline test/tests/smoke_test.rs:141]: The rest LGTM. I'll come back when this is implemented.

I also think it would be fine to open an issue and take care of this in a follow-up

**Review by jhand2 — APPROVED** [review]

File: test/src/lib.rs [hunk 5]
```diff
@@ -1 +1,432 @@
...
+    pub layer: Option<u32>,
+    pub index: Option<u32>,
+    pub fwids: Vec<DiceFwid>,
+    pub flags: Option<u32>,
+    pub vendor_info: Option<Vec<u8>>,
+    pub ty: Option<Vec<u8>>,
+}
+
+impl DiceTcbInfo {
+    fn parse(d: &mut asn1::Parser) -> Result<Self, asn1::ParseError> {
```
Comment (jhand2) [inline test/src/lib.rs:265]: It would probably be easier to use the asn1 crate's parsing trait. See https://github.com/chipsalliance/caliptra-dpe/blob/7ad2c990bf9e06a1374bb2a02a8289761ee8e97f/dpe/src/x509.rs#LL965C1-L965C1

Comment (korran) [reply test/src/lib.rs]: I started down that path originally, but didn't like that I couldn't use ownership types (string, vec, etc) as the field types.

File: test/src/derive.rs [hunk 6]
```diff
@@ -1,5 +1,11 @@
 // Licensed under the Apache-2.0 license
 
+// The IV fed to the DOE when the ROM deobfuscates the UDS seed (as passed to doe registers)
+pub const DOE_UDS_IV: [u32; 4] = [0xfb10365b, 0xa1179741, 0xfba193a1, 0x0f406d7e];
+
+// The IV fed to the DOE when the ROM deobfuscates the field entropy seed (as passed to doe registers)
```
Comment (bluegate010) [inline test/src/derive.rs:25]: Uber nit: no need to refer to this as "field entropy seed". UDS seed was chosen for DICE compliance reasons IIRC.

File: test/src/x509.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    pub vendor: Option<String>,
+    pub model: Option<String>,
+    pub version: Option<String>,
+    pub svn: Option<u32>,
+    pub layer: Option<u32>,
+    pub index: Option<u32>,
+    pub fwids: Vec<DiceFwid>,
+    pub flags: Option<u32>,
+    pub vendor_info: Option<Vec<u8>>,
+    pub ty: Option<Vec<u8>>,
```
Comment (bluegate010) [inline test/src/x509.rs:30]: I presume this is shortened because `type` is a keyword - could this be `type_`?

**Review by bluegate010 — COMMENTED** [review]
Fantastic stuff.

File: test/src/derive.rs (hunk 6 above)
Comment (korran) [reply test/src/derive.rs:25]: This code deals with both the obfuscated field entropy and the deobfuscated field entropy. I've been calling the deobfuscated value the "field entropy" and the obfuscated the "field entropy seed", to match the UDS stuff. What are the correct terms to use to distinctly refer to these two values?

File: test/src/x509.rs (hunk 7 above)
Comment (korran) [reply test/src/x509.rs:30]: Yes. type_ might be a better option, but I've been using ty for this in quite a few other places in the codebase (mostly in ureg and systemrdl libraries).

Comment (bluegate010) [reply test/src/x509.rs:30]: Guess we can change it in a follow-up PR.

File: test/src/derive.rs (hunk 6 above)
Comment (bluegate010) [reply test/src/derive.rs:25]: Ah, I suppose "field entropy seed" is alright then.

**Review by bluegate010 — DISMISSED** [review]

File: test/tests/smoke_test.rs [hunk 8]
```diff
@@ -12,18 +20,133 @@ fn assert_output_contains(haystack: &str, needle: &str) {
...
+    let csr = openssl::x509::X509Req::from_der(&csr_der).unwrap();
+    let csr_txt = String::from_utf8(csr.to_text().unwrap()).unwrap();
+
+    // To update the CSR testdata:
+    // std::fs::write("tests/smoke_testdata/idevid_csr.txt", &csr_txt).unwrap();
+    // std::fs::write("tests/smoke_testdata/idevid_csr.der", &csr_der).unwrap();
+
+    println!("csr: {}", csr_txt);
+
+    assert_eq!(
```
Comment (bluegate010) [inline test/tests/smoke_test.rs:51]: Would it make sense to put some helpful panic message here, like "If your change is expected to modify generated certs or public keys, please re-generate the testdata (see comment in this file)"?

File: test/tests/smoke_test.rs [hunk 9]
```diff
@@ -50,4 +173,115 @@ fn smoke_test() {
...
+        .unwrap()
+        .unwrap();
+    let ldev_cert = openssl::x509::X509::from_der(&ldev_cert_der).unwrap();
+    let ldev_cert_txt = String::from_utf8(ldev_cert.to_text().unwrap()).unwrap();
+
+    // To update the ldev cert testdata:
+    // std::fs::write("tests/smoke_testdata/ldevid_cert.txt", &ldev_cert_txt).unwrap();
+    // std::fs::write("tests/smoke_testdata/ldevid_cert.der", &ldev_cert_der).unwrap();
+
+    assert_eq!(
```
Comment (bluegate010) [inline test/tests/smoke_test.rs:191]: Ditto

**Review by bluegate010 — APPROVED** [review]

File: test/tests/smoke_test.rs [hunk 10]
```diff
@@ -50,4 +157,78 @@ fn smoke_test() {
...
+        "fmc-alias cert: {}",
+        String::from_utf8_lossy(&fmc_alias_cert.to_text().unwrap())
+    );
+
+    let expected_fmc_alias_key = FmcAliasKey::derive(
+        &Pcr0::derive(&Pcr0Input {
+            security_state,
+            fuse_anti_rollback_disable: false,
+            vendor_pub_key_hash: Default::default(),
+            // TODO: Is this right? Should this really be mixed in even if the fuses aren't set?
```
Comment (jhand2) [inline test/tests/smoke_test.rs]: I don't see any harm in doing so. Seems less error prone to unconditionally measure it, even if the value is all 0 or all 0xff.

File: test/src/lib.rs (hunk 5 above)
Comment (jhand2) [reply test/src/lib.rs]: Could just implement Asn1Writable and Asn1Readable for the types you want to use, right? And can use the asn1 functions to implement them, so I'd think they would be small implementations. But I could be wrong.

If it's not too much work, I think it will make things a lot easier/more readable to be able to use the annotations for optional (implicit/explicit) fields.

**Review by jhand2 — APPROVED** [review]

File: test/src/lib.rs (hunk 5 above)
Comment (korran) [reply test/src/lib.rs]: Rust doesn't let you implement an external trait (like AsnReadable) for an external type (like String, Vec). Either the trait or type need to be in my crate.

File: test/tests/smoke_test.rs (hunk 10 above)
Comment (korran) [reply test/tests/smoke_test.rs]: How will we know if the owner fuses are set correctly if we're not measuring the fuse value?

When the owner fuses aren't set, we don't check the signature.

File: test/tests/smoke_test.rs (hunk 8 above)
Comment (korran) [reply test/tests/smoke_test.rs:51]: Yes. Will do this in a followup PR.

File: test/tests/smoke_test.rs (hunk 10 above)
Comment (korran) [reply test/tests/smoke_test.rs]: Filed #415
